### PR TITLE
Add plan button integration 

### PR DESCRIPTION
**Pull request by darvid7**
This PR adds a button at the bottom of the SpecList to add SDK generation plans & connects it with the client to actually create it.

screen shot 2018-07-15 at 1 09 12 am

### CODE REVIEW

File: src/view/basic/BuildStatusChip.tsx [hunk 1]
```diff
@@ -28,6 +28,8 @@ const buildStatusToColor = ({ buildStatus }) => {
       return green[600];
     case BuildStatus.Fail:
       return red[900];
+    default:
```
Comment (darvid7) [inline src/view/basic/BuildStatusChip.tsx]: So I managed to add something to the backend with a null BuildStatus. 
This shouldn't have happened & was removed in commit faec33d618b33cfc9d28621716d7d95f49f322c1 @calebj0seph I don't think it should have been removed.

File: src/state/PlanState.tsx [hunk 2]
```diff
@@ -17,16 +18,28 @@ class PlanState {
...
+    this.plans.set(plan.id, plan);
+  }
 }
 
 export interface AddedPlan {
+  specId: number | null;
   target: string;
   version?: string;
   options?: any;
+  buildStatus: BuildStatus;
```
Comment (PatrickShaw) [inline src/state/PlanState.tsx]: Why is this here? The backend is the source of truth for the build status of a given plan. I don't think the frontend should be dictating what the build status of a plan is.

File: src/view/AddPlanModal.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * A modal window that allows the user to add a SDK Plan to the dashboard.
+ * Currently only supports specifying a target, version and options as a valid JSON.
+ */
+export class AddPlanModal extends Component<RouteComponentProps<{}>, {}> {
+  /**
+   * Whether or not a progress indicator should be shown
+   */
+  @observable private showProgressIndicator: boolean = false;
```
Comment (PatrickShaw) [inline src/view/AddPlanModal.tsx]: All of this kind of looks like the AddSdkModal. Is there anything we can reuse?

File: src/view/Overview.tsx [hunk 4]
```diff
@@ -44,8 +47,15 @@ export class Overview extends Component<RouteComponentProps<{}>, {}> {
               onSpecExpanded={this.expandSpec}
               // Go to the Spec viewing route when you select a Spec
               onSpecSelected={this.goToSpec}
+              //
```
Comment (PatrickShaw) [inline src/view/Overview.tsx]: Is there a reason for this being here?

File: src/view/Overview.tsx [hunk 5]
```diff
@@ -44,8 +47,15 @@ export class Overview extends Component<RouteComponentProps<{}>, {}> {
...
+              onAddSdkPlan={this.addSdkPlan}
             />
             <Route exact path={`${this.props.match.url}/add`} component={AddSpecModal} />
+            <Route
+              exact
+              path={`${this.props.match.url}/plan/add`}
+              component={AddPlanModal}
+            />
```
Comment (PatrickShaw) [inline src/view/Overview.tsx]: I see more similarities between this and the AddSpecModal. Just seems like a lot of boiler plate code. I haven't given much thought into avoiding this issue thus far though.

File: src/view/basic/BuildStatusChip.tsx (hunk 1 above)
Comment (PatrickShaw) [reply src/view/basic/BuildStatusChip.tsx]: Yeah the build status should never be null, at the very least the backend should assign an initial build status to a new plan before it gets added to the database. 

I believe the end game is for builds to be totally separate from plans. You can run a plan as many times as you want and it generates a bunch of 'builds'. You could find out how the latest build is going, for instance, based off this very rough pseudocode: 
```tsx
function latestBuildStatus(planId: Id): Build {
    const buildsCursor: Cursor = query(`SELECT * FROM BUILDS WHERE planId == ${planId} ORDER BY dateCreated DESC LIMIT 1`);
    if (buildsCursor.hasNext()) {
        return buildsCursor.next().buildStatus;
     } else {
        return BuildStatus.NEVER_RUN;
     }
}
```

Comment (PatrickShaw) [reply src/view/basic/BuildStatusChip.tsx]: @darvid7 @calebj0seph @hferrier @TimGebert ^ Let me know if that's not how you picture a plan working.

File: src/view/basic/PlanModal.tsx [hunk 6]
```diff
@@ -177,7 +179,10 @@ export class PlanModal extends Component<PlanModalProps> {
     const target = this.formText.target;
     const version = this.formText.version;
     const options = JSON.parse(this.formText.options);
-    this.props.onSubmitPlan({ target, version, options });
+    const specId = specificationsState.expandedSpecId;
+    const buildStatus = BuildStatus.NotRun;
```
Comment (PatrickShaw) [inline src/view/basic/PlanModal.tsx]: Again, I don't think buildStatus is something that the frontend needs to worry about when adding a plan to swagger-platform. I mean, yes the initial state almost certainly going to be BuildStatus.NotRun, but let the backend decide that for you when it returns a response from `app.service('plans').create(plan)`. At least for now.

File: src/view/basic/BuildStatusChip.tsx (hunk 1 above)
Comment (darvid7) [reply src/view/basic/BuildStatusChip.tsx]: So builds will be stored in `sdks` right? we should probably call them the same thing otherwise confusion. 

The backend assigned the buildStatus to NEVER_RUN before commit faec33d

File: src/state/PlanState.tsx (hunk 2 above)
Comment (darvid7) [reply src/state/PlanState.tsx]: I agree, happy to move this line back then? https://github.com/telstra/swagger-platform/commit/faec33d618b33cfc9d28621716d7d95f49f322c1#diff-47c22a19acf860c06bb7cfa772e82438L104

File: src/view/Overview.tsx (hunk 4 above)
Comment (darvid7) [reply src/view/Overview.tsx]: Nope LOL

File: src/view/basic/PlanModal.tsx (hunk 6 above)
Comment (darvid7) [reply src/view/basic/PlanModal.tsx]: Yea I generally agree, but we removed it from the backend soo had to chuck this in there. I'll move it back

File: src/view/Overview.tsx (hunk 5 above)
Comment (darvid7) [reply src/view/Overview.tsx]: Yea fair enough. I'm not too sure how to go about fixing it though.
I tried to do some stuff https://github.com/darvid7/swagger-platform/tree/add-plan-dodgy/src/view by abstracting the AddPlan & AddSpec away from an AddModal but then errors need to be tied in. I think that can be done if you pass all deps/props in `Overview.tsx` and define `onSubmit()` there but then everything put in that class which doesn't seem like a gr8 idea? 

Not sure if its worth bothering about this now.

File: src/model/Plan.tsx [hunk 7]
```diff
@@ -43,6 +43,21 @@ export enum BuildStatus {
   Fail = 'FAIL',
 }
 
+export const hasValidBuildStatus = buildStatus => {
+  switch (buildStatus) {
+    case BuildStatus.NotRun:
```
Comment (calebj0seph) [inline src/model/Plan.tsx]: Don't use a switch statement, it just duplicates the work when we want to change the enum values. You can simply do:
```javascript
export const hasValidBuildStatus = buildStatus => Object.values(BuildStatus).includes(buildStatus);
```

File: src/view/Overview.tsx [hunk 8]
```diff
@@ -18,7 +19,9 @@ import { AddSpecModal } from 'view/AddSpecModal';
 export class Overview extends Component<RouteComponentProps<{}>, {}> {
   private onSearch = event => {};
   private addSpec = () => this.props.history.push(`${this.props.match.url}/add`);
+  private addSdkPlan = () => this.props.history.push(`${this.props.match.url}/plan/add`);
   private goToSpec = spec => this.props.history.push(`/specs/${spec.id}`);
+  // Don't think this works.
```
Comment (calebj0seph) [inline src/view/Overview.tsx]: What do you mean it doesn't work? Expanding/hiding specs seems to work fine.

File: src/view/Overview.tsx (hunk 4 above)
Comment (calebj0seph) [reply src/view/Overview.tsx]: Might want to add a proper comment then, like 'Open a modal to add a plan when the 'Add Plan' button is clicked'

File: src/view/basic/BuildStatusChip.tsx (hunk 1 above)
Comment (calebj0seph) [reply src/view/basic/BuildStatusChip.tsx]: Yeah I think selecting the build status based of the most recent build is the way to do @PatrickShaw

Comment (calebj0seph) [reply src/view/basic/BuildStatusChip.tsx]: Also yeah might be worth renaming an 'SDK' to a 'build' @darvid7 (although maybe not in this PR)

File: src/view/basic/PlanModal.tsx [hunk 9]
```diff
@@ -173,11 +174,13 @@ export class PlanModal extends Component<PlanModalProps> {
       return;
     }
 
-    // Send the request to the backend
+    // Send the request to the backend.
     const target = this.formText.target;
     const version = this.formText.version;
     const options = JSON.parse(this.formText.options);
-    this.props.onSubmitPlan({ target, version, options });
+    const specId = specificationsState.expandedSpecId;
```
Comment (calebj0seph) [inline src/view/basic/PlanModal.tsx]: Seems a bit hacky to assume that the spec ID is just the currently expanded spec. I think it'd be better to pass the spec ID in as a prop to the modal to make it more reusable.

File: src/view/basic/SpecList.tsx [hunk 10]
```diff
@@ -11,6 +11,7 @@ export interface SpecListProps extends React.DOMAttributes<HTMLDivElement> {
   expandedSpecId: number | null;
   onSpecExpanded: (id: number | null) => void;
   onSpecSelected: (Spec: Spec) => void;
+  onAddSdkPlan: () => void;
```
Comment (calebj0seph) [inline src/view/basic/SpecList.tsx]: I feel like this and `addSpec` in Overview.tsx should be renamed to something like `openAddPlanModal` and `openAddSpecModal` respectively for clarity.

**Comment by PatrickShaw** [comment]
@darvid7 can you add export `storybook` thing? I couldn't see one for the Component. Preferably `AddSdkPlanModal` would show up in storybook.

File: src/view/basic/BuildStatusChip.tsx (hunk 1 above)
Comment (darvid7) [reply src/view/basic/BuildStatusChip.tsx]: yea, we should probs sort our naming out cuz I still think plan is kinda weird, mb we should ask Steve but worry about it later imo

File: src/model/Plan.tsx (hunk 7 above)
Comment (darvid7) [reply src/model/Plan.tsx]: tyvm

File: src/view/Overview.tsx (hunk 8 above)
Comment (darvid7) [reply src/view/Overview.tsx]: Yea idk what I was thinking, ignore that

File: src/view/basic/SpecList.tsx (hunk 10 above)
Comment (darvid7) [reply src/view/basic/SpecList.tsx]: 👍 I called the plan stuff `openAddSdkPlanModal` because on the UI it says `Add Sdk`. I don't think users will understand what a plan is.

File: src/view/basic/PlanModal.tsx (hunk 9 above)
Comment (darvid7) [reply src/view/basic/PlanModal.tsx]: It would have to be the currently expanded spec though wouldn't it?
You can only expand one spec at a time. I don't see any other way to add a Plan because it has to be for a Spec so you need to have clicked on the Spec first?

If you pass it into the props wouldn't it propagate from `Overview.tsx` so it would just be `  private expandSpec = id => (state.expandedSpecId = id);`?

**Comment by darvid7** [comment]
@PatrickShaw Not sure how to do that, I think `PlanModal.tsx` is already in storybook but not `AddPlanModal.tsx`. Link to docs pls

**Comment by PatrickShaw** [comment]
Yep my mistake, I think it's fine, PlanModal is the one with the storybook, AddPlanModal doesn't need one.

**Comment by darvid7** [comment]
@calebj0seph  ty for the fixes
